torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -222,13 +221,25 @@ def foreach_reduce_scatter_copy_in(
 
 def _get_all_gather_input_metadatas(
     param_all_gather_inputs: List[List[torch.Tensor]],
-) -> Tuple[List[List[torch.dtype]], List[List[int]]]:
+) -> Tuple[List[List[torch.dtype]], List[List[int]], torch.dtype]:
     param_all_gather_input_dtypes: List[List[torch.dtype]] = []
     param_all_gather_input_numels: List[List[int]] = []
+    all_gather_dtype = param_all_gather_inputs[0][0].dtype
     for all_gather_inputs in param_all_gather_inputs:
-        param_all_gather_input_dtypes.append([t.dtype for t in all_gather_inputs])
-        param_all_gather_input_numels.append([t.numel() for t in all_gather_inputs])
-    return param_all_gather_input_dtypes, param_all_gather_input_numels
+        input_dtypes: List[torch.dtype] = []
+        input_numels: List[int] = []
+        for all_gather_input in all_gather_inputs:
+            if all_gather_input.dtype != all_gather_dtype:
+                all_gather_dtype = torch.uint8
```

Comment:
so this is handling the case where a FSDPParamGroup consists of different tensor dtypes, where when we perform allgather on this group, we just treat them universally as uint8 and recover their dtypes later?

Reply:
yep!